:has(~:dir(~)) should work

rniwa opened the pull request:
#### ded49fab22337a0f9d321bc1b13e5ddeb49c917d
<pre>
:has(~:dir(~)) should work
<a href="https://bugs.webkit.org/show_bug.cgi?id=243093">https://bugs.webkit.org/show_bug.cgi?id=243093</a>

Reviewed by Antti Koivisto.

Implement :dir pseudo class invalidation. This patch introduces the notion of effective text direction
which is stored in Node&apos;s rareDataBitfields. To signify whether this is used for text direction or not,
another boolean value usesEffectiveTextDirection has been added to rareDataBitfields.

Like <a href="https://commits.webkit.org/253764@main">https://commits.webkit.org/253764@main</a> for :lang, this patch has an optimization to avoid using
rareDataBitfields when the text direction matches that of the document element.

* LayoutTests/TestExpectations: Removed image only failures from tests that are now passing.
* LayoutTests/fast/css/dir-pseudo-container-query-expected.html: Added.
* LayoutTests/fast/css/dir-pseudo-container-query-invalidation-expected.html: Added.
* LayoutTests/fast/css/dir-pseudo-container-query-invalidation.html: Added.
* LayoutTests/fast/css/dir-pseudo-container-query.html: Added.
* LayoutTests/platform/win/TestExpectations: Skip the newly added tests since :dir isn&apos;t enabled yet.

* Source/WebCore/css/SelectorCheckerTestFunctions.h
(WebCore::matchesDirPseudoClass): Now uses effectiveTextDirection instead of computeDirectionality.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::childrenChanged): Calls setDocumentElementTextDirection. Note we need to explicitly
check usesEffectiveTextDirection since effectiveTextDirection falls back to documentElementTextDirection.

* Source/WebCore/dom/Document.h:
(WebCore::Document::documentElementTextDirection const): Added.
(WebCore::Document::setDocumentElementTextDirection): Added.

* Source/WebCore/dom/Node.cpp:
(WebCore::Node::incrementConnectedSubframeCount):
(WebCore::Node::effectiveTextDirection const): Added.
(WebCore::Node::setEffectiveTextDirection): Added.
(WebCore::Node::setUsesEffectiveTextDirection): Added.

* Source/WebCore/dom/Node.h:
(WebCore::Node::usesEffectiveTextDirection const): Added.

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::TextDirectionDirective): Added.
(WebCore::parseTextDirection): Added.
(WebCore::isValidDirValue): Now uses parseTextDirection.
(WebCore::elementAffectsDirectionality): Moved up to be used in insertedIntoAncestor.
(WebCore::HTMLElement::insertedIntoAncestor): Inherit the effective text direction from parent unless
this is an input type=telephone.
(WebCore::HTMLElement::removedFromAncestor): Added. Clear the inherited effective text direction.
(WebCore::HTMLElement::computeDirectionality const): Removed.
(WebCore::HTMLElement::dirAttributeChanged): Added the invalidation logic for each value type. Notably,
when the new value is invalid, we inherit from the parent node unless this is an input type=telephone.
(WebCore::HTMLElement::updateEffectiveDirectionality): Added. Implements the recursive invalidation
for :dir pseudo class.
(WebCore::HTMLElement::adjustDirectionalityIfNeededAfterChildAttributeChanged): Update the effective
text direction instead of simply invalidating the style.
(WebCore::HTMLElement::updateEffectiveDirectionalityOfDirAuto): Ditto.
(WebCore::HTMLElement::adjustDirectionalityIfNeededAfterChildrenChanged):

* Source/WebCore/html/HTMLElement.h:

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::subtreeHasChanged): Don&apos;t call updateEffectiveDirectionalityOfDirAuto
if this element hasn&apos;t been marked with selfOrPrecedingNodesAffectDirAuto. Otherwise, we&apos;d set
selfOrPrecedingNodesAffectDirAuto on input element without dir=auto. This was caught by new assertions.
(WebCore::HTMLInputElement::setValue): Added a call to updateEffectiveDirectionalityOfDirAuto when
selfOrPrecedingNodesAffectDirAuto is set. This is needed now that the effective text direction is cached
as opposed to re-computed each time in computeDirectionality.

* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::subtreeHasChanged): Same as HTMLInputElement.
(WebCore::HTMLTextAreaElement::setValueCommon): Ditto.

Canonical link: <a href="https://commits.webkit.org/254017@main">https://commits.webkit.org/254017@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/760a9df3d1fc3d7e7d9e88905b81ad9d66513ab6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87750 "failed 4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31859 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18482 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96950 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/150758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91718 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30203 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/26288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79855 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91693 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93369 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24399 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74495 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24382 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79357 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67229 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27890 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13341 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27863 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14356 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2824 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29549 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37269 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29447 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33632 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on LayoutTests/TestExpectations:
```diff
@@ -1579,9 +1579,6 @@ webkit.org/b/185859 imported/w3c/web-platform-tests/css/selectors/focus-visible-
-webkit.org/b/64861 imported/w3c/web-platform-tests/css/selectors/dir-selector-auto-direction-change-001.html [ ImageOnlyFailure ]
-webkit.org/b/64861 imported/w3c/web-platform-tests/css/selectors/dir-selector-change-001.html [ ImageOnlyFailure ]
-webkit.org/b/64861 imported/w3c/web-platform-tests/css/selectors/dir-selector-change-004.html [ ImageOnlyFailure ]
 webkit.org/b/64861 imported/w3c/web-platform-tests/css/selectors/dir-style-01a.html [ ImageOnlyFailure ]
 webkit.org/b/64861 imported/w3c/web-platform-tests/css/selectors/dir-style-01b.html [ ImageOnlyFailure ]
```
wonder what is still wrong with these?

rniwa replied:
It's anti-aliasing issue. See https://github.com/WebKit/WebKit/pull/3842

rniwa committed:
:has(~:dir(~)) should work
https://bugs.webkit.org/show_bug.cgi?id=243093

Reviewed by Antti Koivisto.

Implement :dir pseudo class invalidation. This patch introduces the notion of effective text direction
which is stored in Node's rareDataBitfields. To signify whether this is used for text direction or not,
another boolean value usesEffectiveTextDirection has been added to rareDataBitfields.

Like https://commits.webkit.org/253764@main for :lang, this patch has an optimization to avoid using
rareDataBitfields when the text direction matches that of the document element.

* LayoutTests/TestExpectations: Removed image only failures from tests that are now passing.
* LayoutTests/fast/css/dir-pseudo-container-query-expected.html: Added.
* LayoutTests/fast/css/dir-pseudo-container-query-invalidation-expected.html: Added.
* LayoutTests/fast/css/dir-pseudo-container-query-invalidation.html: Added.
* LayoutTests/fast/css/dir-pseudo-container-query.html: Added.
* LayoutTests/platform/win/TestExpectations: Skip the newly added tests since :dir isn't enabled yet.

* Source/WebCore/css/SelectorCheckerTestFunctions.h
(WebCore::matchesDirPseudoClass): Now uses effectiveTextDirection instead of computeDirectionality.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::childrenChanged): Calls setDocumentElementTextDirection. Note we need to explicitly
check usesEffectiveTextDirection since effectiveTextDirection falls back to documentElementTextDirection.

* Source/WebCore/dom/Document.h:
(WebCore::Document::documentElementTextDirection const): Added.
(WebCore::Document::setDocumentElementTextDirection): Added.

* Source/WebCore/dom/Node.cpp:
(WebCore::Node::incrementConnectedSubframeCount):
(WebCore::Node::effectiveTextDirection const): Added.
(WebCore::Node::setEffectiveTextDirection): Added.
(WebCore::Node::setUsesEffectiveTextDirection): Added.

* Source/WebCore/dom/Node.h:
(WebCore::Node::usesEffectiveTextDirection const): Added.

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::TextDirectionDirective): Added.
(WebCore::parseTextDirection): Added.
(WebCore::isValidDirValue): Now uses parseTextDirection.
(WebCore::elementAffectsDirectionality): Moved up to be used in insertedIntoAncestor.
(WebCore::HTMLElement::insertedIntoAncestor): Inherit the effective text direction from parent unless
this is an input type=telephone.
(WebCore::HTMLElement::removedFromAncestor): Added. Clear the inherited effective text direction.
(WebCore::HTMLElement::computeDirectionality const): Removed.
(WebCore::HTMLElement::dirAttributeChanged): Added the invalidation logic for each value type. Notably,
when the new value is invalid, we inherit from the parent node unless this is an input type=telephone.
(WebCore::HTMLElement::updateEffectiveDirectionality): Added. Implements the recursive invalidation
for :dir pseudo class.
(WebCore::HTMLElement::adjustDirectionalityIfNeededAfterChildAttributeChanged): Update the effective
text direction instead of simply invalidating the style.
(WebCore::HTMLElement::updateEffectiveDirectionalityOfDirAuto): Ditto.
(WebCore::HTMLElement::adjustDirectionalityIfNeededAfterChildrenChanged):

* Source/WebCore/html/HTMLElement.h:

* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::HTMLInputElement::subtreeHasChanged): Don't call updateEffectiveDirectionalityOfDirAuto
if this element hasn't been marked with selfOrPrecedingNodesAffectDirAuto. Otherwise, we'd set
selfOrPrecedingNodesAffectDirAuto on input element without dir=auto. This was caught by new assertions.
(WebCore::HTMLInputElement::setValue): Added a call to updateEffectiveDirectionalityOfDirAuto when
selfOrPrecedingNodesAffectDirAuto is set. This is needed now that the effective text direction is cached
as opposed to re-computed each time in computeDirectionality.

* Source/WebCore/html/HTMLTextAreaElement.cpp:
(WebCore::HTMLTextAreaElement::subtreeHasChanged): Same as HTMLInputElement.
(WebCore::HTMLTextAreaElement::setValueCommon): Ditto.

Canonical link: https://commits.webkit.org/254017@main
